tmk_core/common/chibios/eeprom_stm32f4.c

```diff
@@ -0,0 +1,378 @@
+ * too big for eeprom emulation
+ */
+#define EE_SECTOR_ID 3
+#define EE_SECTOR_SIZE 0x4000
+#define EE_SECTOR_BEGIN 0x0800C000
+#define EE_SECTOR_END (EE_SECTOR_BEGIN+EE_SECTOR_SIZE)
+#define EE_MAX_BYTES 1024 //compatible with avr which has 1024 bytes eeprom
+#define IS_VALID_FLASH(ADDRESS) (((ADDRESS) >= EE_SECTOR_BEGIN) && ((ADDRESS) < EE_SECTOR_END))
+#define IS_VALID_ADDRESS(ADDRESS) (((ADDRESS) >= 0) && ((ADDRESS) < EE_MAX_BYTES))
+#define DATA_ADDRESS(d) (((d)>>16)&0xFFFF)
+#define DATA_VALUE(d) (((d)&0xFFFF))
+#define DATA_COMPOSE(a,d) (((a)<<16)|(d))
+
+#define FLASH_PSIZE_BYTE 0
+#define FLASH_PSIZE_HFWORD FLASH_CR_PSIZE_0
+#define FLASH_PSIZE_WORD FLASH_CR_PSIZE_1
+#define FLASH_CR_SNB_POS 3
+/* the flash key was not defined in the current CMSIS */
+#define FLASH_KEY1_F4 0x45670123U
+#define FLASH_KEY2_F4 0xCDEF89ABU
+#define FLASH_EMPTY_VALUE 0xFFFFFFFFU
+
+/* backup buffer */
+static uint8_t DataBuf[EE_MAX_BYTES];
+
+static FLASH_Status FLASH_StatusF4(void) {
+    if (FLASH->SR & FLASH_SR_BSY) return FLASH_BUSY;
+
+    if ((FLASH->SR & (FLASH_SR_PGSERR | FLASH_SR_PGPERR | FLASH_SR_PGAERR))) return FLASH_ERROR_PG;
+
+    if ((FLASH->SR & FLASH_SR_WRPERR)) return FLASH_ERROR_WRP;
+
+    return FLASH_COMPLETE;
+}
+
+static void FLASH_DelayF4(void) {
+    __IO uint32_t i = 0;
+    for (i = 0xFF; i != 0; i--) {
+        __asm__ volatile("nop\n\t");
+    }
+}
+
+static FLASH_Status FLASH_WaitF4(void) {
+    /* Check for the Flash Status */
+    FLASH_Status status = FLASH_StatusF4();
+    /* Wait for a Flash operation to complete */
+    while( status == FLASH_BUSY) {
+        FLASH_DelayF4();
+        status = FLASH_StatusF4();
+    }
+
+    return status;
+}
+
+static void FLASH_ClearFlagF4(void) {
+    if (FLASH->SR & FLASH_SR_PGAERR) {
+        FLASH->SR |= FLASH_SR_PGAERR;
+    }
+    if (FLASH->SR & FLASH_SR_PGPERR) {
+        FLASH->SR |= FLASH_SR_PGPERR;
+    }
+    if (FLASH->SR & FLASH_SR_PGSERR) {
+        FLASH->SR |= FLASH_SR_PGSERR;
+    }
+    if (FLASH->SR & FLASH_SR_WRPERR) {
+        FLASH->SR |= FLASH_SR_WRPERR;
+    }
+}
+
+static void FLASH_UnlockF4(void) {
+    FLASH->KEYR = FLASH_KEY1_F4;
+    FLASH->KEYR = FLASH_KEY2_F4;
+}
+
+static FLASH_Status FLASH_EraseSectorF4(uint32_t sector) {
+    FLASH_Status status = FLASH_WaitF4();
+    if (status == FLASH_COMPLETE) {
+        /* if the previous operation is completed, proceed to erase the page */
+        FLASH->CR &= ~FLASH_CR_SNB;
+        FLASH->CR |= (sector << FLASH_CR_SNB_POS);
+        FLASH->CR |= FLASH_CR_SER;
+        FLASH->CR |= FLASH_CR_STRT;
+
+        /* Wait for last operation to be completed */
+        status = FLASH_WaitF4();
+        /* clear the erase bit */
+        FLASH->CR &= ~FLASH_CR_SNB;
+        FLASH->CR &= ~FLASH_CR_SER;
+        if (status != FLASH_COMPLETE) {
+            FLASH_ClearFlagF4();
+        }
+    }
+    /* Return the Erase Status */
+    return status;
+}
```

Comment:
Conventional erased value for an external flash is 0xFF for each byte -- is the erased value after this operation 0x00 or 0xFF?
eeconfig will not function correctly if the default value is non-zero after an erase.

Reply:
I suppose writing 0xFF for everything means the iteration through the list won't actually match anything when querying values. In hindsight, this should be fine.

Reply:
After flash was erased, all the bit will be set to "1". We can only program the bit from "1" to "0", and can't from "0" to "1" which need a full page erasing operation. Thus the 0xFF means the empty value in flash.